[dtensor][3/n] use dedicated TensorMeta instead of the fx one

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109145
* #109144
* #108264
* #108263
* #108262
* __->__ #108261
* #107306

This PR switches the usage of fx's shape prop TensorMetadata to
dtensor's own dedicated defined TensorMeta, this is because DTensor
only cares three fields: shape/stride/dtype, all other fields are not
necessary and can be inferred from local_tensor directly. This would
help significantly simplify how we deal with the tensor metadata by not
caring other fields.

wanchaol committed:
[dtensor][3/n] use dedicated TensorMeta instead of the fx one

This PR switches the usage of fx's shape prop TensorMetadata to
dtensor's own dedicated defined TensorMeta, this is because DTensor
only cares three fields: shape/stride/dtype, all other fields are not
necessary and can be inferred from local_tensor directly. This would
help significantly simplify how we deal with the tensor metadata by not
caring other fields.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][3/n] use dedicated TensorMeta instead of the fx one"


This PR switches the usage of fx's shape prop TensorMetadata to
dtensor's own dedicated defined TensorMeta, this is because DTensor
only cares three fields: shape/stride/dtype, all other fields are not
necessary and can be inferred from local_tensor directly. This would
help significantly simplify how we deal with the tensor metadata by not
caring other fields.

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -367,14 +366,23 @@ def __str__(self) -> str:
         return "P"
 
 
+class TensorMeta(NamedTuple):
```
Not a strong opinion here, maybe call it `DTensorMeta`? WDYT?

wanchaol replied:
I think for this dataclass is intended to record "tensor" metadata like shape/stride/dtype, these fields are not quite related to `DTensor`, if we rename it to "DTensorMeta" I think we would expect this to include "device_mesh" and "placements", which is already recorded in the DTensorSpec, this might lead to some confusion. I'll also think more about the naming and see whether we need to consolidate it with DTensorSpec a bit

fduwjj reviewed (approved):
Other than the naming, the rest looks good to me.

wanchaol committed:
Update on "[dtensor][3/n] use dedicated TensorMeta instead of the fx one"


This PR switches the usage of fx's shape prop TensorMetadata to
dtensor's own dedicated defined TensorMeta, this is because DTensor
only cares three fields: shape/stride/dtype, all other fields are not
necessary and can be inferred from local_tensor directly. This would
help significantly simplify how we deal with the tensor metadata by not
caring other fields.

[ghstack-poisoned]
